File: rest-service/src/main/java/mb/lib/query/BlastValidationManager.java
package mb.lib.query;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.*;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import mb.api.service.model.ErrorMap;
import mb.api.service.util.Address;
import mb.lib.config.Config;
import mb.lib.util.JSON;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.veupathdb.lib.blast.BlastConfig;
import org.veupathdb.lib.container.jaxrs.errors.UnprocessableEntityException;

class BlastValidationManager
{
  private static final Logger Log  = LogManager.getLogger(BlastValidationManager.class);
  private static final Config Conf = Config.getInstance();
  private static final String Path = "validate";

  public static Optional<ErrorMap> Validate(BlastConfig conf)
  throws Exception {
    Log.trace("::Validate(conf={})", conf);

    var payload = JSON.stringify(conf);
    var uri     = Address.http(String.join("/", Conf.getValidatorHost(), Path, conf.getTool().getValue()));
    Log.debug("Sending config to {}: {}", uri, payload);

    var res = HttpClient.newHttpClient()
      .send(
        HttpRequest.newBuilder()
          .POST(HttpRequest.BodyPublishers.ofString(payload))
          .uri(URI.create(uri))
          .build(),
        HttpResponse.BodyHandlers.ofString()
      );

    if (res.statusCode() != 200)
      throw new RuntimeException("Unexpected response from validation server: " + res.body());

    var parsed = JSON.parse(res.body(), ValidationResponse.class);

    Log.debug("JSON Reply: {}", res.body());
    if (parsed.payload.isEmpty())
      return Optional.empty();

    var converted = new ErrorMap(parsed.payload.size());

    for (var key : parsed.payload.keySet()) {
      var flag = Field2Flag.Flag2Field.get(key);

      if (flag == null)
        continue;

      converted.put(flag, parsed.payload.get(key));
    }

    return Optional.of(converted);
  }
}

class ValidationResponse {
  final int status;
  final Map<String, List<String>> payload;

  @JsonCreator
  public ValidationResponse(
    @JsonProperty("status")  int                       status,
    @JsonProperty("payload") Map<String, List<String>> payload
  ) {
    this.status  = status;
    this.payload = payload == null ? Collections.emptyMap() : payload;
  }
}

File: rest-service/src/main/java/mb/lib/query/BlastValidationManager.kt
package mb.lib.query

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import mb.api.service.model.ErrorMap
import mb.api.service.util.Address
import mb.lib.config.Config
import mb.lib.util.jsonStringify
import mb.lib.util.parseJSON
import org.apache.logging.log4j.LogManager
import org.veupathdb.lib.blast.BlastConfig
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.util.*

object BlastValidationManager {
  private val Log  = LogManager.getLogger(BlastValidationManager::class)

  private const val Path = "validate"

  fun validate(conf: BlastConfig): ErrorMap? {
    Log.trace("::Validate(conf={})", conf)

    val payload = conf.jsonStringify()
    val uri     = Address.http(java.lang.String.join("/", Config.validatorHost, Path, conf.tool.value))
    Log.debug("Sending config to {}: {}", uri, payload)

    val res = HttpClient.newHttpClient()
      .send(
        HttpRequest.newBuilder()
          .POST(HttpRequest.BodyPublishers.ofString(payload))
          .uri(URI.create(uri))
          .build(),
        HttpResponse.BodyHandlers.ofString()
      )

    if (res.statusCode() != 200)
      throw RuntimeException("Unexpected response from validation server: " + res.body())

    val parsed = res.body().parseJSON<ValidationResponse>()

    Log.debug("JSON Reply: {}", res.body())
    if (parsed.payload.isEmpty())
      return null

    val converted = ErrorMap(parsed.payload.size)

    for ((key, value) in parsed.payload) {
      val flag = Field2Flag.Flag2Field[key] ?: continue

      converted[flag] = value
    }

    return converted
  }
}

private data class ValidationResponse(
  @JsonProperty("status")  val status: Int,
  @JsonProperty("payload") val payload: Map<String, MutableList<String>> = emptyMap()
)